File: src/main/java/Decorator/AbstractFactory.java
package Decorator;


interface Factory {
    abstract ICar create();
}

class MercedesFactory implements Factory {
    @Override
    public ICar create() {
        return new Mercedes();
    }
}

class RenaultFactory implements Factory {
    @Override
    public ICar create() {
        return new Renault();
    }
}

class BmwFactory implements Factory {
    @Override
    public ICar create() {
        return new Bmw();
    }
}

File: src/main/kotlin/AbstractFactory.kt
interface CarFactory {
    fun create(): Car
}

class MercedesFactory : CarFactory {
    override fun create(): Car = Mercedes()
}

class PeogeotFactory : CarFactory {
    override fun create(): Car = Peogeot()
}

class RenaultFactory : CarFactory {
    override fun create(): Car = Renault()
}

interface Car
class Mercedes : Car
class Renault : Car
class Peogeot : Car